README.rst
=====
WIMMS
=====

**WIMMS** stands for *Where Is My Mozilla Service ?*.

This library implements a SQL backend for a database that
lists for a list of service, (user, node, service) tuples.

It's used by Mozilla's Token Server.

To see where WIMMS fits in the whole architecture, see:

.. image:: http://ziade.org/token-org.png

Tests
-----

To run the tests for mysql, you need to install a mysqlserver and export the 
'WIMMS_MYSQLURI' variable properly. Something like this::

    $ export WIMMS_MYSQLURI="mysql://user:pass@host/db"


CHANGES.rst
CHANGES
=======

XXXX - 0.4
----------

- Removed "service acceptance" flags and the related metadata table.
- Added tracking of optional "generation number" for each user account,
  which can be used to more quickly detect password change events.
- Added tracking of optional "client state" string for each user account,
  which can be used to force node reallocation when client state changes.

2012-07-24 - 0.3
----------------

- Added a metadata table which is able to store information related to the user
  and service.

2012-07-18 - 0.2.1
------------------

- allocated_node will just return an existing node.


2012-04-18 - 0.2
----------------

- added the pool_reset_on_return option


2012-04-17 - 0.1
----------------

- initial release


wimms/tests/test_sql.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this file,
# You can obtain one at http://mozilla.org/MPL/2.0/.
from unittest2 import TestCase
import os
import uuid
from mozsvc.exceptions import BackendError
from wimms.sql import SQLMetadata


TEMP_ID = uuid.uuid4().hex


class NodeAssignmentTests(object):

    backend = None  # subclasses must define this on the instance

    def setUp(self):
        super(NodeAssignmentTests, self).setUp()
        self.backend.add_service('sync-1.0', '{node}/1.0/{uid}')
        self.backend.add_service('sync-1.5', '{node}/1.5/{uid}')
        self.backend.add_service('queuey-1.0', '{node}/{service}/{uid}')
        self.backend.add_node('sync-1.0', 'https://phx12', 100)

    def test_node_allocation(self):
        user = self.backend.get_user("sync-1.0", "tarek@mozilla.com")
        self.assertEquals(user, None)

        user = self.backend.create_user("sync-1.0", "tarek@mozilla.com")
        wanted = 'https://phx12'
        self.assertEqual(user['node'], wanted)

        user = self.backend.get_user("sync-1.0", "tarek@mozilla.com")
        self.assertEqual(user['node'], wanted)

    def test_allocation_to_least_loaded_node(self):
        self.backend.add_node('sync-1.0', 'https://phx13', 100)
        user1 = self.backend.create_user("sync-1.0", "test1@mozilla.com")
        user2 = self.backend.create_user("sync-1.0", "test2@mozilla.com")
        self.assertNotEqual(user1['node'], user2['node'])

    def test_update_generation_number(self):
        user = self.backend.create_user("sync-1.0", "tarek@mozilla.com")
        self.assertEqual(user['generation'], 0)
        self.assertEqual(user['client_state'], '')
        orig_uid = user['uid']
        orig_node = user['node']

        # Changing generation should leave other properties unchanged.
        self.backend.update_user("sync-1.0", user, generation=42)
        self.assertEqual(user['uid'], orig_uid)
        self.assertEqual(user['node'], orig_node)
        self.assertEqual(user['generation'], 42)
        self.assertEqual(user['client_state'], '')

        user = self.backend.get_user("sync-1.0", "tarek@mozilla.com")
        self.assertEqual(user['uid'], orig_uid)
        self.assertEqual(user['node'], orig_node)
        self.assertEqual(user['generation'], 42)
        self.assertEqual(user['client_state'], '')

        # It's not possible to move generation number backwards.
        self.backend.update_user("sync-1.0", user, generation=17)
        self.assertEqual(user['uid'], orig_uid)
        self.assertEqual(user['node'], orig_node)
        self.assertEqual(user['generation'], 42)
        self.assertEqual(user['client_state'], '')

        user = self.backend.get_user("sync-1.0", "tarek@mozilla.com")
        self.assertEqual(user['uid'], orig_uid)
        self.assertEqual(user['node'], orig_node)
        self.assertEqual(user['generation'], 42)
        self.assertEqual(user['client_state'], '')

    def test_update_client_state(self):
        user = self.backend.create_user("sync-1.0", "tarek@mozilla.com")
        self.assertEqual(user['generation'], 0)
        self.assertEqual(user['client_state'], '')
        self.assertEqual(set(user['old_client_states']), set(()))
        seen_uids = set((user['uid'],))
        orig_node = user['node']

        # Changing client-state allocates a new userid.
        self.backend.update_user("sync-1.0", user, client_state="aaa")
        self.assertTrue(user['uid'] not in seen_uids)
        self.assertEqual(user['node'], orig_node)
        self.assertEqual(user['generation'], 0)
        self.assertEqual(user['client_state'], 'aaa')
        self.assertEqual(set(user['old_client_states']), set(("",)))

        user = self.backend.get_user("sync-1.0", "tarek@mozilla.com")
        self.assertTrue(user['uid'] not in seen_uids)
        self.assertEqual(user['node'], orig_node)
        self.assertEqual(user['generation'], 0)
        self.assertEqual(user['client_state'], 'aaa')
        self.assertEqual(set(user['old_client_states']), set(("",)))

        seen_uids.add(user['uid'])

        # It's possible to change client-state and generation at once.
        self.backend.update_user("sync-1.0", user,
                                 client_state="bbb", generation=12)
        self.assertTrue(user['uid'] not in seen_uids)
        self.assertEqual(user['node'], orig_node)
        self.assertEqual(user['generation'], 12)
        self.assertEqual(user['client_state'], 'bbb')
        self.assertEqual(set(user['old_client_states']), set(("", "aaa")))

        user = self.backend.get_user("sync-1.0", "tarek@mozilla.com")
        self.assertTrue(user['uid'] not in seen_uids)
        self.assertEqual(user['node'], orig_node)
        self.assertEqual(user['generation'], 12)
        self.assertEqual(user['client_state'], 'bbb')
        self.assertEqual(set(user['old_client_states']), set(("", "aaa")))

        # You can't got back to an old client_state.
        orig_uid = user['uid']
        with self.assertRaises(BackendError):
            self.backend.update_user("sync-1.0", user, client_state="aaa")

        user = self.backend.get_user("sync-1.0", "tarek@mozilla.com")
        self.assertEqual(user['uid'], orig_uid)
        self.assertEqual(user['node'], orig_node)
        self.assertEqual(user['generation'], 12)
        self.assertEqual(user['client_state'], 'bbb')
        self.assertEqual(set(user['old_client_states']), set(("", "aaa")))


class TestSQLDB(NodeAssignmentTests, TestCase):

    _SQLURI = os.environ.get('WIMMS_SQLURI', 'sqlite:////tmp/wimms.' + TEMP_ID)

    def setUp(self):
        self.backend = SQLMetadata(self._SQLURI, create_tables=True)
        super(TestSQLDB, self).setUp()

    def tearDown(self):
        super(TestSQLDB, self).tearDown()
        if self.backend._engine.driver == 'pysqlite':
            filename = self.backend.sqluri.split('sqlite://')[-1]
            if os.path.exists(filename):
                os.remove(filename)
        else:
            self.backend._safe_execute('drop table services;')
            self.backend._safe_execute('drop table nodes;')
            self.backend._safe_execute('drop table users;')


if os.environ.get('WIMMS_MYSQLURI', None) is not None:
    class TestMySQLDB(TestSQLDB):
        _SQLURI = os.environ.get('WIMMS_MYSQLURI')


wimms/tests/test_shardedsql.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this file,
# You can obtain one at http://mozilla.org/MPL/2.0/.
import os
from unittest2 import TestCase

from wimms.shardedsql import ShardedSQLMetadata, ENGINE_INDEX
from wimms.tests.test_sql import NodeAssignmentTests, TEMP_ID


_SQLURI = os.environ.get('WIMMS_SQLURI', 'sqlite:////tmp/wimms.' + TEMP_ID)
_SQLURI = 'sync-1.0;%s,queuey;%s' % (_SQLURI, _SQLURI)


class TestSQLShardedDB(NodeAssignmentTests, TestCase):

    def setUp(self):
        self.backend = ShardedSQLMetadata(_SQLURI, create_tables=True)
        super(TestSQLShardedDB, self).setUp()

    def tearDown(self):
        for service, value in self.backend._dbs.items():
            engine = value[ENGINE_INDEX]
            sqlite = engine.driver == 'pysqlite'
            if sqlite:
                filename = str(engine.url).split('sqlite://')[-1]
                if os.path.exists(filename):
                    os.remove(filename)
            else:
                engine.execute('drop table services')
                engine.execute('drop table nodes')
                engine.execute('drop table users')


wimms/schemas.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this file,
# You can obtain one at http://mozilla.org/MPL/2.0/.
"""
Table schema for MySQL and sqlite.

We have the following tables:

 services:  lists the available services and their endpoint-url pattern.
 nodes:  lists the nodes available for each service.
 users:  lists the user records for each service, along with their
         metadata and current node assignment.

"""

from sqlalchemy.ext.declarative import declared_attr
from sqlalchemy import Column, Integer, String, BigInteger, Index


bases = {}


def _add(name, base):
    bases[name] = base


def get_cls(name, base_cls):
    if name in base_cls.metadata.tables:
        return base_cls.metadata.tables[name]

    args = {'__tablename__': name}
    base = bases[name]
    return type(name, (base, base_cls), args).__table__


class _UsersBase(object):
    """This table associates email addresses with services via integer uids.

    A user is uniquely identified by their email.  For each service they have
    a uid, an allocated node, and last-seen generation and client-state values.
    Rows are timestamped for easy cleanup of old records.
    """
    uid = Column(BigInteger(), primary_key=True, autoincrement=True,
                 nullable=False)
    service = Column(Integer(), nullable=False)
    email = Column(String(255), nullable=False)
    node = Column(String(64), nullable=False)
    generation = Column(BigInteger(), nullable=False)
    client_state = Column(String(32), nullable=False)
    created_at = Column(BigInteger(), nullable=False)
    replaced_at = Column(BigInteger(), nullable=True)

    @declared_attr
    def __table_args__(cls):
        return (
            # Index used to slurp in all records for a (service, email)
            # pair, sorted by creation time.
            Index('lookup_idx', 'email', 'service', 'created_at'),
            # Index to prevent duplicate records for a single
            # client_state value.
            Index('clientstate_idx', 'email', 'service', 'client_state',
                  unique=True),
            # Index used for purging user_records that have been replaced.
            Index('replaced_at_idx', 'service', 'replaced_at'),
            {'mysql_engine': 'InnoDB', 'mysql_charset': 'utf8'}
        )

_add('users', _UsersBase)


class _ServicesBase(object):
    """This table lists all the available services and their endpoint patterns.

    Service names are expected to be "{app_name}-{app_version}" for example
    "sync-1.5".  Endpoint patterns can use python formatting options on the
    keys {uid}, {node} and {service}.

    Having a table for these means that we can internally refer to each service
    by an integer key, which helps when indexing by service.
    """
    id = Column(Integer(), primary_key=True, autoincrement=True,
                nullable=False)
    service = Column(String(30), unique=True)
    pattern = Column(String(128))

_add('services', _ServicesBase)


class _NodesBase(object):
    """This table keeps tracks of all nodes available per service

    Each node has a root URL as well as metadata about its current availability
    and capacity.
    """
    id = Column(BigInteger(), primary_key=True, autoincrement=True,
                nullable=False)
    service = Column(Integer(), nullable=False)
    node = Column(String(64), nullable=False)
    available = Column(Integer, default=0, nullable=False)
    current_load = Column(Integer, default=0, nullable=False)
    capacity = Column(Integer, default=0, nullable=False)
    downed = Column(Integer, default=0, nullable=False)
    backoff = Column(Integer, default=0, nullable=False)

    @declared_attr
    def __table_args__(cls):
        return (
            Index('unique_idx', 'service', 'node', unique=True),
            {'mysql_engine': 'InnoDB', 'mysql_charset': 'utf8'}
        )

_add('nodes', _NodesBase)


wimms/shardedsql.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this file,
# You can obtain one at http://mozilla.org/MPL/2.0/.
"""
Shareded Service Metadata Database.

This implementation provides the same interface as SQLMetadata, but uses
a separate database for each service.  This can help with managing extremely
high load, by keeping the sizes of each table smaller.
"""
from mozsvc.exceptions import BackendError

from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import create_engine
from sqlalchemy.pool import NullPool
from sqlalchemy.sql import select

from wimms.sql import SQLMetadata

ENGINE_INDEX = 0
SERVICES_INDEX = 1
NODES_INDEX = 2
USERS_INDEX = 3


class ShardedSQLMetadata(SQLMetadata):

    def __init__(self, databases, create_tables=False, pool_size=100,
                 pool_recycle=60, pool_timeout=30, max_overflow=10,
                 pool_reset_on_return='rollback', **kw):

        self._cached_service_ids = {}
        # databases is a string containing one sqluri per service:
        #   service1;sqluri1,service2;sqluri2
        self._dbs = {}
        if pool_reset_on_return.lower() in ('', 'none'):
            pool_reset_on_return = None

        for database in databases.split(','):
            database = database.split(';')
            service, sqluri = (el.strip() for el in database)
            if self._dbkey(service) in self._dbs:
                continue

            Base = declarative_base()
            if sqluri.startswith('mysql') or sqluri.startswith('pymysql'):
                engine = create_engine(
                    sqluri,
                    pool_size=pool_size,
                    pool_recycle=pool_recycle,
                    pool_timeout=pool_timeout,
                    max_overflow=max_overflow,
                    pool_reset_on_return=pool_reset_on_return,
                    logging_name='wimms'
                )
            else:
                # XXX will use a shared pool next
                engine = create_engine(sqluri, poolclass=NullPool)

            engine.echo = kw.get('echo', False)

            self._is_sqlite = (engine.driver == 'pysqlite')
            if self._is_sqlite:
                from wimms.sqliteschemas import get_cls  # NOQA
            else:
                from wimms.schemas import get_cls   # NOQA

            services = get_cls('services', Base)
            nodes = get_cls('nodes', Base)
            users = get_cls('users', Base)

            for table in (services, nodes, users):
                table.metadata.bind = engine
                if create_tables:
                    table.create(checkfirst=True)

            self._dbs[self._dbkey(service)] = (engine, services, nodes, users)

    def _dbkey(self, service):
        """Strip version number, returning just the service name."""
        return service.split('-')[0]

    def _get_engine(self, service=None):
        if service is None:
            raise NotImplementedError()
            return self._dbs.values()[0][ENGINE_INDEX]
        return self._dbs[self._dbkey(service)][ENGINE_INDEX]

    def _get_table(self, service, index):
        return self._dbs[self._dbkey(service)][index]

    def _get_services_table(self, service):
        return self._get_table(service, SERVICES_INDEX)

    def _get_nodes_table(self, service):
        return self._get_table(service, NODES_INDEX)

    def _get_users_table(self, service):
        return self._get_table(service, USERS_INDEX)

    def get_patterns(self):
        """Returns all the service URL patterns."""
        # loop on all the tables to combine the pattern information.
        patterns = []
        for service, elements in self._dbs.items():
            engine = elements[0]
            table = elements[SERVICES_INDEX]
            try:
                res = self._safe_execute(select([table]), engine=engine)
            except BackendError:
                continue
            try:
                for row in res:
                    self._cached_service_ids[row.service] = row.id
                    if row not in patterns:
                        patterns.append(row)
            finally:
                res.close()
        return patterns

    def add_service(self, service, pattern):
        """Add definition for a new service."""
        engine = self._get_engine(service)
        return super(ShardedSQLMetadata, self).add_service(service, pattern,
                                                           engine=engine)


wimms/sqliteschema.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this file,
# You can obtain one at http://mozilla.org/MPL/2.0/.
"""
    Table schema for MySQL and sqlite
"""
import traceback
from mozsvc.exceptions import BackendError

from sqlalchemy.sql import select, update, and_
from sqlalchemy.ext.declarative import declarative_base, declared_attr, Column
from sqlalchemy import Integer, String, create_engine, BigInteger, Index
from sqlalchemy.pool import NullPool
from sqlalchemy.sql import text as sqltext
from sqlalchemy.exc import OperationalError, TimeoutError

from wimms import logger


_Base = declarative_base()
tables = []


def get_user_nodes_table(driver, base=_Base):
    if 'user_nodes' in _Base.metadata.tables:
        return _Base.metadata.tables['user_nodes']

    if driver != 'pysqlite':
        class UserNodes(base):
            """This table lists all the users associated to a service.

            A user is represented by an email, a uid and its allocated node.
            """
            __tablename__ = 'user_nodes'
            email = Column(String(255), nullable=False)
            node = Column(String(64), nullable=False)
            service = Column(String(30), nullable=False)
            uid = Column(BigInteger(), primary_key=True, autoincrement=True,
                         nullable=False)

            __table_args__ = (Index('userlookup_idx',
                                    'email', 'service', unique=True),
                              Index('nodelookup_idx',
                                    'node', 'service'),
                              {'mysql_engine': 'InnoDB',
                               'mysql_charset': 'utf8',
                              },
                             )

        return UserNodes.__table__
    else:

        class UserNodes(base):
            """Sqlite version"""
            __tablename__ = 'user_nodes'
            email = Column(String(255))
            node = Column(String(64), nullable=False)
            service = Column(String(30))
            uid = Column(Integer(11), primary_key=True, autoincrement=True)

        return UserNodes.__table__


class _ServicePatternBase(object):
    """ A table that keeps track of the url pattern.
    """
    service = Column(String(30), primary_key=True)
    pattern = Column(String(128), primary_key=True)


class ServicePattern(_ServicePatternBase, _Base):
    __tablename__ = 'service_pattern'


service_pattern = ServicePattern.__table__
tables.append(service_pattern)


class _NodesBase(object):
    """A Table that keep tracks of all nodes per service
    """
    id = Column(BigInteger(), primary_key=True, autoincrement=True,
                nullable=False)
    service = Column(String(30), nullable=False)
    node = Column(String(64), nullable=False)
    available = Column(Integer, default=0, nullable=False)
    current_load = Column(Integer, default=0, nullable=False)
    capacity = Column(Integer, default=0, nullable=False)
    downed = Column(Integer, default=0, nullable=False)
    backoff = Column(Integer, default=0, nullable=False)

    @declared_attr
    def __table_args__(cls):
        return (Index('unique_idx', 'service', 'node', unique=True),
                {'mysql_engine': 'InnoDB'},
               )

class Nodes(_NodesBase, _Base):
    __tablename__ = 'nodes'


nodes = Nodes.__table__
tables.append(nodes)




wimms/sql.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this file,
# You can obtain one at http://mozilla.org/MPL/2.0/.
"""
Service Metadata Database.

For each available service, we maintain a list of user accounts and their
associated uid, node-assignment and metadata.  We also have a list of nodes
with their load, capacity etc
"""
import time
import traceback
from mozsvc.exceptions import BackendError

from sqlalchemy.sql import select, update, and_
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import create_engine
from sqlalchemy.pool import NullPool
from sqlalchemy.sql import text as sqltext, func as sqlfunc
from sqlalchemy.exc import OperationalError, TimeoutError, IntegrityError

from wimms import logger


def get_timestamp():
    """Get current timestamp in milliseconds."""
    return int(time.time() * 1000)


_Base = declarative_base()


_GET_USER_RECORDS = sqltext("""\
select
    uid, node, generation, client_state
from
    users
where
    email = :email
and
    service = :service
order by
    created_at desc, uid desc
limit
    20
""")


_CREATE_USER_RECORD = sqltext("""\
insert into
    users
    (service, email, node, generation, client_state, created_at, replaced_at)
values
    (:service, :email, :node, :generation, :client_state, :timestamp, NULL)
""")


_UPDATE_GENERATION_NUMBER = sqltext("""\
update
    users
set
    generation = :generation
where
    service = :service and email = :email and
    generation < :generation and replaced_at is null
""")


_REPLACE_USER_RECORDS = sqltext("""\
update
    users
set
    replaced_at = :timestamp
where
    service = :service and email = :email
    and replaced_at is null and created_at < :timestamp
""")


WRITEABLE_FIELDS = ['available', 'current_load', 'capacity', 'downed',
                    'backoff']


class SQLMetadata(object):

    def __init__(self, sqluri, create_tables=False, pool_size=100,
                 pool_recycle=60, pool_timeout=30, max_overflow=10,
                 pool_reset_on_return='rollback', **kw):
        self._cached_service_ids = {}
        self.sqluri = sqluri
        if pool_reset_on_return.lower() in ('', 'none'):
            pool_reset_on_return = None

        if sqluri.startswith('mysql') or sqluri.startswith('pymysql'):
            self._engine = create_engine(
                sqluri,
                pool_size=pool_size,
                pool_recycle=pool_recycle,
                pool_timeout=pool_timeout,
                pool_reset_on_return=pool_reset_on_return,
                max_overflow=max_overflow,
                logging_name='wimms'
            )
        else:
            self._engine = create_engine(sqluri, poolclass=NullPool)

        self._engine.echo = kw.get('echo', False)

        self._is_sqlite = (self._engine.driver == 'pysqlite')
        if self._is_sqlite:
            from wimms.sqliteschemas import get_cls  # NOQA
        else:
            from wimms.schemas import get_cls  # NOQA

        self.services = get_cls('services', _Base)
        self.nodes = get_cls('nodes', _Base)
        self.users = get_cls('users', _Base)

        for table in (self.services, self.nodes, self.users):
            table.metadata.bind = self._engine
            if create_tables:
                table.create(checkfirst=True)

    def _get_engine(self, service=None):
        return self._engine

    def _safe_execute(self, *args, **kwds):
        """Execute an sqlalchemy query, raise BackendError on failure."""
        if hasattr(args[0], 'bind'):
            engine = args[0].bind
        else:
            engine = None

        if engine is None:
            engine = kwds.pop('engine', None)
            if engine is None:
                engine = self._get_engine(kwds.get('service'))

        if 'service' in kwds:
            kwds['service'] = self._get_service_id(kwds['service'])

        try:
            return engine.execute(*args, **kwds)
        except (OperationalError, TimeoutError), exc:
            err = traceback.format_exc()
            logger.error(err)
            raise BackendError(str(exc))

    def get_user(self, service, email):
        params = {'service': service, 'email': email}
        res = self._safe_execute(_GET_USER_RECORDS, **params)
        try:
            row = res.fetchone()
            if row is None:
                return None
            # The first row is the most up-to-date user record.
            user = {
                'email': email,
                'uid': row.uid,
                'node': row.node,
                'generation': row.generation,
                'client_state': row.client_state,
                'old_client_states': {}
            }
            # Any subsequent rows are due to old client-state values.
            row = res.fetchone()
            while row is not None:
                user['old_client_states'][row.client_state] = True
                row = res.fetchone()
            return user
        finally:
            res.close()

    def create_user(self, service, email, generation=0, client_state=''):
        node = self.get_best_node(service)
        params = {
            'service': service, 'email': email, 'node': node,
            'generation': generation, 'client_state': client_state,
            'timestamp': get_timestamp()
        }
        try:
            res = self._safe_execute(_CREATE_USER_RECORD, **params)
        except IntegrityError:
            return self.get_user(service, email)
        else:
            res.close()
            return {
                'email': email,
                'uid': res.lastrowid,
                'node': node,
                'generation': generation,
                'client_state': client_state,
                'old_client_states': {}
            }

    def update_user(self, service, user, generation=None, client_state=None):
        if client_state is None:
            # uid can stay the same, just update the generation number.
            if generation is not None:
                params = {
                    'service': service,
                    'email': user['email'],
                    'generation': generation
                }
                res = self._safe_execute(_UPDATE_GENERATION_NUMBER, **params)
                res.close()
                user['generation'] = max(generation, user['generation'])
        else:
            # reject previously-seen client-state strings.
            if client_state == user['client_state']:
                raise BackendError('previously seen client-state string')
            if client_state in user['old_client_states']:
                raise BackendError('previously seen client-state string')
            # need to create a new record for new client_state.
            if generation is not None:
                generation = max(user['generation'], generation)
            else:
                generation = user['generation']
            now = get_timestamp()
            params = {
                'service': service, 'email': user['email'],
                'node': user['node'], 'timestamp': now,
                'generation': generation, 'client_state': client_state
            }
            try:
                res = self._safe_execute(_CREATE_USER_RECORD, **params)
            except IntegrityError:
                user.update(self.get_user(service, user['email']))
            else:
                self.get_user(service, user['email'])
                user['uid'] = res.lastrowid
                user['generation'] = generation
                user['old_client_states'][user['client_state']] = True
                user['client_state'] = client_state
                res.close()
            # mark old records as having been replaced.
            # if we crash here, they are unmarked and we may fail to
            # garbage collect them for a while, but the active state
            # will be undamaged.
            params = {
                'service': service, 'email': user['email'], 'timestamp': now
            }
            res = self._safe_execute(_REPLACE_USER_RECORDS, **params)
            res.close()

    #
    # Nodes management
    #

    def _get_service_id(self, service):
        try:
            return self._cached_service_ids[service]
        except KeyError:
            services = self._get_services_table(service)
            query = select([services.c.id])
            query = query.where(services.c.service == service)
            res = self._safe_execute(query)
            row = res.fetchone()
            res.close()
            if row is None:
                raise BackendError('unknown service: ' + service)
            self._cached_service_ids[service] = row.id
            return row.id

    def get_patterns(self):
        """Returns all the service URL patterns."""
        query = select([self.services])
        res = self._safe_execute(query)
        patterns = list(res.fetchall())
        for row in patterns:
            self._cached_service_ids[row.service] = row.id
        res.close()
        return patterns

    def add_service(self, service, pattern, **kwds):
        """Add definition for a new service."""
        res = self._safe_execute(sqltext("""
          insert into services (service, pattern)
          values (:servicename, :pattern)
        """), servicename=service, pattern=pattern, **kwds)
        res.close()
        return res.lastrowid

    def add_node(self, service, node, capacity, **kwds):
        """Add definition for a new node."""
        res = self._safe_execute(sqltext(
            """
            insert into nodes (service, node, available, capacity,
                               current_load, downed, backoff)
            values (:service, :node, :available, :capacity,
                    :current_load, :downed, :backoff)
            """),
            service=service, node=node, capacity=capacity,
            available=kwds.get('available', capacity),
            current_load=kwds.get('current_load', 0),
            downed=kwds.get('downed', 0),
            backoff=kwds.get('backoff', 0),
        )
        res.close()

    def get_best_node(self, service):
        """Returns the 'least loaded' node currently available, increments the
        active count on that node, and decrements the slots currently available
        """
        nodes = self._get_nodes_table(service)
        service = self._get_service_id(service)

        where = [nodes.c.service == service,
                 nodes.c.available > 0,
                 nodes.c.capacity > nodes.c.current_load,
                 nodes.c.downed == 0]

        query = select([nodes]).where(and_(*where))

        if self._is_sqlite:
            # sqlite doesn't have the 'log' funtion, and requires
            # coercion to a float for the sorting to work.
            query = query.order_by(nodes.c.current_load * 1.0 /
                                   nodes.c.capacity)
        else:
            # using log() increases floating-point precision on mysql
            # and thus makes the sorting more accurate.
            query = query.order_by(sqlfunc.log(nodes.c.current_load) /
                                   sqlfunc.log(nodes.c.capacity))
        query = query.limit(1)
        res = self._safe_execute(query)
        one = res.fetchone()
        if one is None:
            # unable to get a node
            res.close()
            raise BackendError('unable to get a node')

        node = str(one.node)
        res.close()

        # updating the table
        where = [nodes.c.service == service, nodes.c.node == node]
        where = and_(*where)
        fields = {'available': nodes.c.available - 1,
                  'current_load': nodes.c.current_load + 1}
        query = update(nodes, where, fields)
        con = self._safe_execute(query, close=True)
        con.close()

        return node

    def _get_services_table(self, service):
        return self.services

    def _get_nodes_table(self, service):
        return self.nodes

    def _get_users_table(self, service):
        return self.users
